src/coverage/coverageService.ts

```diff
@@ -0,0 +1,163 @@
+// Copyright 2019 Espressif Systems (Shanghai) CO LTD
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { basename, join } from "path";
+import * as vscode from "vscode";
+import { appendIdfAndToolsToPath, execChildProcess } from "../utils";
+import { OutputChannel } from "../logger/outputChannel";
+import { Logger } from "../logger/logger";
+
+export interface countRange {
+  range: vscode.Range;
+  count: Number;
+}
+
+export interface textEditorWithCoverage {
+  coveredLines: vscode.Range[];
+  partialLines: vscode.Range[];
+  uncoveredLines: vscode.Range[];
+  editor: vscode.TextEditor;
+  countPerLines: countRange[];
+  allLines: vscode.Range[];
+}
+
+export async function buildJson(dirPath: string) {
+  const result = await _runCmd(
+    "gcovr",
```

Comment:
Also directly calling of `gcovr` is not launching even though `pip list` have 
`gcovr           4.2` version !!
(_I am using python 2.7 btw_)

Reply:
Should add to  python virtual env in `idf.customExtraPaths` in https://github.com/espressif/vscode-esp-idf-extension/pull/104

Reply:
👍